### PR TITLE
Feature transformation

### PR DESCRIPTION
**Pull request by eleenkmail**
- Do initial Transformation on data :
    
   - One hot encoding for categorical features 
   - Extract day, month and year from dates features 
   - Scaling data with: 
   -- standard scaler
   -- min-max scaler

### CODE REVIEW

File: src/utils/transform_data.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# import libraries
+import pandas as pd
+from sklearn.preprocessing import OneHotEncoder
+
+
```
Comment (zubaidasader) [inline src/utils/transform_data.py:5]: add docString to each function to match the code style for all functions.

**Review by zubaidasader — APPROVED** [review]

File: src/utils/__init__.py [hunk 2]
```diff
@@ -1 +1,3 @@
-from .general_look_at_data import *
\ No newline at end of file
+from transform_data import *
```
Comment (mawada-sweis) [inline src/utils/__init__.py:1]: This code line motivated me to look for the difference between those lines, which is the point, and this caused me to discover a better approach to call the function than the one we currently use! Thank you, Eleen.
The technique will be updated later, but it consists of removing the dot, adding the path line, and replacing the module name with the actual path. We'll talk about it at our next meeting.

File: notebooks/Transformation.ipynb [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+   "source": [
+    "date_features = ['first_travel_date','last_travel_date']  #date features \n",
+    "categorical_features = ['cms_zone', 'work_cms_zone']  #categorical features\n",
+    "\n",
+    "# all features to be encoded include categorical and labeled features\n",
+    "features_to_encode = ['cms_zone', 'work_cms_zone','smartphone_type','made_trips', 'employment',\n",
+    "'age', 'student','vehicle',  'congestion','drive_cbd_freq', 'planning_apps','job_type', 'industry',\n",
+    "'telework_freq', 'work_mode','bike_freq', 'tnc_freq', 'tnc_purpose',\n",
+    "'tnc_mode', 'packages', 'harassment', 'gender', 'education', 'english_proficiency', 'ethnicity']"
+   ]
```
Comment (Rama-Has) [inline notebooks/Transformation.ipynb:90]: I prefer to use label encoder instead of the one-hot encoder for the following features: english_proficiency, education, tnc_fre, telework_freq, planning_apps, drive_cbd_freq, congestion and age

**Review by Rama-Has — APPROVED** [review]

File: notebooks/Transformation.ipynb [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+  },
```
[the remark this answers is not in the document]
Comment (eleenkmail) [reply notebooks/Transformation.ipynb:91]: yes we can ,but we must reverse the order in for these features:
tnc_fre, telework_freq, planning_apps, drive_cbd_freq, congestion, english_proficiency

[the remark this answers is not in the document]
Comment (Rama-Has) [reply notebooks/Transformation.ipynb:2464]: We could use mapping to maintain the required order of data

[the remark this answers is not in the document]
Comment (Rama-Has) [reply notebooks/Transformation.ipynb:2464]: + to reverse the needed values

[the remark this answers is not in the document]
Comment (eleenkmail) [reply notebooks/Transformation.ipynb:2464]: we have 3 types of missing 
1-3?
or 0-2?
then 3/4 to categories-1 ?

[the remark this answers is not in the document]
Comment (eleenkmail) [reply notebooks/Transformation.ipynb:2464]: what do you think is better?

[the remark this answers is not in the document]
Comment (eleenkmail) [reply notebooks/Transformation.ipynb:2464]: I think we do not have to do reverse label encoding since the algorithm depends on the distance the arrange reversed or not it is not important

and the numbers are arranged, so we don't need to do label encoding either, we just want to do the mapping to replace missing codes with other values in the same arrange.

**Review by eleenkmail — COMMENTED** [review]
Done

File: src/utils/general_look_at_data.py [hunk 5]
```diff
@@ -70,12 +70,9 @@ def display_direct_missing(df):
...
     # Filter rows with missing percentage greater than 0
     filtered_stats = df[df.drop(columns='feature_type').gt(0).any(axis=1)]
-    
-    # Get indirect missing columns with at least one feature above 0.0
-    indirect_missing_cols = filtered_stats.drop(columns=['missing_count', 'missing_percentage'])
-    indirect_missing_cols = indirect_missing_cols.columns[indirect_missing_cols.any()]
+
     # Return indirect missing columns
-    return filtered_stats[indirect_missing_cols]
+    return filtered_stats.drop(columns=['missing_count', 'missing_percentage'])
```
Comment (mawada-sweis) [inline src/utils/general_look_at_data.py:75]: In this case, you will return all features regardless of the percentage of missing... previous code returns the columns that have at least one feature above 0.0...

File: src/utils/general_look_at_data.py [hunk 6]
```diff
@@ -85,19 +82,13 @@ def plot_indirect_missing(missing_data):
...
-    if num_features == 0:
-        print("No features found with missing data")
-        return
+    num_features = len(filtered_data.columns)
     num_cols = 2
     num_rows = math.ceil(num_features / num_cols)
     fig, axs = plt.subplots(num_rows, num_cols, figsize=(10, 5*num_rows))
 
-    for i, column_name in enumerate(non_zero_features):
+    for i, column_name in enumerate(filtered_data.columns):
```
Comment (mawada-sweis) [inline src/utils/general_look_at_data.py:91]: why did you update this code?

Comment (eleenkmail) [reply src/utils/general_look_at_data.py:91]: Not me 😃

Comment (eleenkmail) [reply src/utils/general_look_at_data.py:91]: Last change on general_look_at_data.py was before merge the Feature transformation pull request to main

**Review by mawada-sweis — APPROVED** [review]
